### PR TITLE
Add report backend for intermediate caching.

### PR DESCRIPTION
**Pull request by tkaemming**
This allows reports to be built once for each project in an organization, so that individual user reports can be primarily composed from cached data.

@getsentry/infrastructure

---

This change is [Reviewable](https://reviewable.io/reviews/getsentry/sentry/3931)

### CODE REVIEW

**Comment by tkaemming** [comment]
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[src/sentry/tasks/reports.py, line 265 \[r1\]](https://reviewable.io:443/reviews/getsentry/sentry/3931#-KP-HupwfwalWtKCVJAz:-KP-HupwfwalWtKCVJB-:b3t3zbc) ([raw file](https://github.com/getsentry/sentry/blob/589a254abf24bac9d33c53fef2eb5da2152e5074/src/sentry/tasks/reports.py#L265)):*
> ```Python
> 
> 
> class DummyReportBackend(ReportBackend):
> ```

This will be used for testing, currently unused.

---

*[src/sentry/tasks/reports.py, line 326 \[r1\]](https://reviewable.io:443/reviews/getsentry/sentry/3931#-KP-H_uRnXwspWW6-WHU:-KP-H_uRnXwspWW6-WHV:b-z6b1zx) ([raw file](https://github.com/getsentry/sentry/blob/589a254abf24bac9d33c53fef2eb5da2152e5074/src/sentry/tasks/reports.py#L326)):*
> ```Python
> 
> 
> backend = RedisReportBackend(
> ```

This will eventually be configurable like most things in `sentry.app`, but I'd like to wait until things shake out before introducing that much flexibility. (This is all still considered private API as far as I'm concerned.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/sentry/3931)*

File: src/sentry/tasks/reports.py [hunk 1]
```diff
@@ -2,8 +2,10 @@
 
 import functools
 import itertools
+import json
```
Comment (mattrobenolt) [inline src/sentry/tasks/reports.py]: Use our `sentry.utils.json` instead. It's a bit more optimized and will be faster.

File: src/sentry/tasks/reports.py [hunk 2]
```diff
@@ -238,6 +241,94 @@ def prepare_project_report(interval, project):
...
+        )
+
+    def __encode(self, report):
+        return zlib.compress(json.dumps(report))
+
+    def __decode(self, value):
+        return json.loads(zlib.decompress(value))
+
+    def prepare(self, timestamp, duration, organization):
+        reports = {}
```
Comment (mattrobenolt) [inline src/sentry/tasks/reports.py]: Will this be problematic if people have multiple hundreds of projects? Definitely not a blocker, but the extreme side is ~1500 projects or so.

**Comment by tkaemming** [comment]
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[src/sentry/tasks/reports.py, line 5 \[r1\]](https://reviewable.io:443/reviews/getsentry/sentry/3931#-KP-G_0nGN3eYUnpmVMa-r1-5:-KP-XaSz1FivNRSgApFn:b-15yaad) ([raw file](https://github.com/getsentry/sentry/blob/589a254abf24bac9d33c53fef2eb5da2152e5074/src/sentry/tasks/reports.py#L5)):*
<details><summary><i>Previously, mattrobenolt (Matt Robenolt) wrote…</i></summary>
> Use our `sentry.utils.json` instead. It's a bit more optimized and will be faster.
</details>
I'll go ahead and make this change, but this scares me a little bit — if a `datetime` ends up in the report in the future, that's going to cause some pretty confusing issues.

---

*[src/sentry/tasks/reports.py, line 305 \[r1\]](https://reviewable.io:443/reviews/getsentry/sentry/3931#-KP-G_0nGN3eYUnpmVMa-r1-305:-KP-YOvLimYkilS8ULzq:ba0b7h2) ([raw file](https://github.com/getsentry/sentry/blob/589a254abf24bac9d33c53fef2eb5da2152e5074/src/sentry/tasks/reports.py#L305)):*
<details><summary><i>Previously, mattrobenolt (Matt Robenolt) wrote…</i></summary>
> Will this be problematic if people have multiple hundreds of projects? Definitely not a blocker, but the extreme side is ~1500 projects or so.
</details>
I can't think of any specific issues. Was there something you had in mind? (For context, a single project report with all items filled is ~100 bytes right now after encoding.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/sentry/3931)*

**Comment by tkaemming** [comment]
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/sentry/tasks/reports.py, line 5 \[r1\]](https://reviewable.io:443/reviews/getsentry/sentry/3931#-KP-G_0nGN3eYUnpmVMa-r1-5:-KP-abmrx1ouo9UbA1OZ:b-896fix) ([raw file](https://github.com/getsentry/sentry/blob/589a254abf24bac9d33c53fef2eb5da2152e5074/src/sentry/tasks/reports.py#L5)):*
<details><summary><i>Previously, tkaemming (ted kaemming) wrote…</i></summary>
> I'll go ahead and make this change, but this scares me a little bit — if a `datetime` ends up in the report in the future, that's going to cause some pretty confusing issues.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/sentry/3931)*

**Comment by mattrobenolt** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/sentry/tasks/reports.py, line 5 \[r1\]](https://reviewable.io:443/reviews/getsentry/sentry/3931#-KP-G_0nGN3eYUnpmVMa-r1-5:-KP-aZnv3gpkh1oxU5Wv:blkeq1n) ([raw file](https://github.com/getsentry/sentry/blob/589a254abf24bac9d33c53fef2eb5da2152e5074/src/sentry/tasks/reports.py#L5)):*
<details><summary><i>Previously, tkaemming (ted kaemming) wrote…</i></summary>
> Done.
</details>
I'm not sure I understand the issue. Our json would encode the datetime object just fine, and the stdlib would explode on it. Unless you mean that you'd want it to explode? Or is the problem there because it's lossy. If you were to encode datetime, it won't decode back into datetime?

Alternatively, could just pickle here and avoid it all. This would be a safe place to pickle.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/sentry/3931)*

**Comment by tkaemming** [comment]
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[src/sentry/tasks/reports.py, line 5 \[r1\]](https://reviewable.io:443/reviews/getsentry/sentry/3931#-KP-G_0nGN3eYUnpmVMa-r1-5:-KP-buCzlI2ZhkE4uKJM:bfrg7t4) ([raw file](https://github.com/getsentry/sentry/blob/589a254abf24bac9d33c53fef2eb5da2152e5074/src/sentry/tasks/reports.py#L5)):*
> Unless you mean that you'd want it to explode? Or is the problem there because it's lossy. If you were to encode datetime, it won't decode back into datetime?

Yeah — for the purposes here, it should either decode the same value it encoded, or refuse to encode at all.

> Alternatively, could just pickle here and avoid it all. This would be a safe place to pickle.

This inflates the payload by about 50% :(

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/sentry/3931)*

**Comment by mattrobenolt** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[src/sentry/tasks/reports.py, line 5 \[r1\]](https://reviewable.io:443/reviews/getsentry/sentry/3931#-KP-G_0nGN3eYUnpmVMa-r1-5:-KP-e2iirm50r1CFHjRE:b-fltvlh) ([raw file](https://github.com/getsentry/sentry/blob/589a254abf24bac9d33c53fef2eb5da2152e5074/src/sentry/tasks/reports.py#L5)):*
<details><summary><i>Previously, tkaemming (ted kaemming) wrote…</i></summary>
> > Unless you mean that you'd want it to explode? Or is the problem there because it's lossy. If you were to encode datetime, it won't decode back into datetime?
> 
> Yeah — for the purposes here, it should either decode the same value it encoded, or refuse to encode at all.
> 
> > Alternatively, could just pickle here and avoid it all. This would be a safe place to pickle.
> 
> This inflates the payload by about 50% :(
</details>
Yeah, this is fine then.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/sentry/3931)*

**Comment by mattrobenolt** [comment]
Reviewed 1 of 1 files at r1.
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/sentry/3931)*

**Comment by mattrobenolt** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[src/sentry/tasks/reports.py, line 305 \[r1\]](https://reviewable.io:443/reviews/getsentry/sentry/3931#-KP-G_0nGN3eYUnpmVMa-r1-305:-KP-eDAWECyK3Nsf_kSS:bey03ec) ([raw file](https://github.com/getsentry/sentry/blob/589a254abf24bac9d33c53fef2eb5da2152e5074/src/sentry/tasks/reports.py#L305)):*
<details><summary><i>Previously, tkaemming (ted kaemming) wrote…</i></summary>
> I can't think of any specific issues. Was there something you had in mind? (For context, a single project report with all items filled is ~100 bytes right now after encoding.)
</details>
Nope, just giving you a heads up of what the bounds are here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/sentry/3931)*

**Comment by mattrobenolt** [comment]
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/sentry/3931)*

**Comment by mattrobenolt** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/sentry/3931)*

**Pull request merged**